adr/2022_03_31_specification_pattern_introduced.md

```diff
@@ -0,0 +1,359 @@
+# Introduction of Specification Pattern
+
+* Status: proposed
+* Date: 2022-03-31
+
+## Context and Problem Statement
+
+During the implementation of Headless Checkout [Pull Request](https://github.com/Sylius/Sylius/pull/13793) we encountered
+need to use a Specification pattern as multiple business requirements must be met before completing checkout.
+
+### What Headless Checkout is?
+The current checkout system is highly coupled with UI, therefore to maintain the current checkout system, but use it via API
+a user had to make at least 5 requests in a specific order (add to cart, address, payment select, shipping select, complete).
+The new concept would reduce the number of requests to the API and get rid of the order requirement.
```

Comment:
```suggestion
The current checkout system is highly coupled with UI. To use it via API, multiple requests must be made in a specific order: add to cart, address, select payment, select shipping, complete.
The new concept would allow completing checkout with fewer requests in any order.
This is crucial for one-page checkouts, where the user might perform any action in any order.
```